### PR TITLE
Basic NeuralFP Implementaiton

### PR DESCRIPTION
**Pull request by tianweiy**
This pull request includes basic support for Neural Fingerprinting. Basic changes are 1. NeuralFP class in a separate file 2. NFLoss in loss function script 3. Two helper methods in the pytorch_utils.

### CODE REVIEW

**Comment by tianweiy** [comment]
@revbucket Can you have a review of my NeuralFP.py program and especially the generation of the data set. Currently I first use the val data_loader with a batch size of 64, and only use the first image of the batch so that I cut the test set size by 64x. There may be some bug in the program as I notice an unreasonable increase of false positive when increasing the threshold(although notice this only once). Do you have any other suggestion to build a small dataset?

**Comment by revbucket** [comment]
Couple of notes: (mostly just me being neurotic, in general this looks like it works just fine on your computer)

**Meat**
1) Nice work on writing down the right signs of the combined loss function. That's probably my least favorite  thing about mister_ed, and I have to think about it every time. 

2) The loss function looks good, as far as replicating the experiments goes. Let's try and get the following loss function, which might do a little better:
``` sum_j Max([Tau - D(x, f, xi^j), 0]) ```
(does that make sense? We don't get any benefit by making a fingerprint distance go _higher_ than Tau, so let's stop optimizing for it)


**Styling**
1) I can't see any bugs offhand from the data generation step, but it's probably easier and cleaner to wrap up any large monolithic subroutines into their own methods. For example, in the dataset generation, having a subroutine to compute the fingerprint loss, which itself calls a subroutine that generates the dataset would make it really easy to debug any problems in dataset generation. See, for example, the _attack_subroutine method

2) Continuing the line of neuroticism: it's probably much cleaner and easier to organize (if not for your sake, at least for the sake of others who'll read your code) all the neuralFP stuff in a directory and then you can have multiple files. 

3) line lengths are a pet peeve of mine. <=80 is preferred, but if it goes over 100, it's really hard for me to read on my laptop screen in two-panel mode.

4) I'd love to run this myself, but the locally referenced files make it impossible. Share the weights? (put them on dropbox, or we can work out a way to share them irl)

**Comment by tianweiy** [comment]
Thanks, please see my reply in the email.
